site/cds_rdm/components.py

```diff
@@ -130,3 +129,100 @@ def publish(self, identity, draft=None, record=None, **kwargs):
+
+    def create(self, identity, data=None, record=None, errors=None, **kwargs):
+        """Mint/update alternative identifiers on create."""
+        alt_id_schemes = current_app.config["CDS_CERN_MINT_ALTERNATE_IDS"]
+
+        alternate_identifiers_map = {}
+        for id in data["metadata"].get("identifiers", []):
+            if id["scheme"] in alt_id_schemes:
+                alternate_identifiers_map.setdefault(id["scheme"], []).append(
+                    id["identifier"]
+                )
+        if not alternate_identifiers_map:
+            # If no mintable identifiers, return early
+            return
+
+        # Bulk query for the record's alternative identifiers
+        existing_pids = PersistentIdentifier.query.filter(
+            PersistentIdentifier.pid_type.in_(alt_id_schemes.keys()),
+            PersistentIdentifier.object_type == "rec",
+            PersistentIdentifier.object_uuid == record.pid.object_uuid,
+        ).all()
+        # Build a set of (scheme, value) pairs that already exist
+        existing_pairs = {(pid.pid_type, pid.pid_value): pid for pid in existing_pids}
+
+        def already_exists_error_message(scheme, value):
+            """Return the error message for a duplicate identifier."""
+            return {
+                "field": "metadata.identifiers",
+                "messages": [
+                    _(
+                        f"Identifier value '{value}' for scheme '{alt_id_schemes[scheme]}' already exists."
+                    )
+                ],
+            }
+
+        for scheme, values in alternate_identifiers_map.items():
+            for value in values:
+                if (scheme, value) in existing_pairs:
+                    pid = existing_pairs[(scheme, value)]
+                    # Remove the already existing ones from the dictionary
+                    del existing_pairs[(scheme, value)]
+                    if pid.object_uuid == record.pid.object_uuid:
+                        # The PID already exists for the current record, so we can skip it
+                        continue
+                    errors.append(already_exists_error_message(scheme, value))
+                else:
+                    try:
+                        PersistentIdentifier.create(
+                            pid_type=scheme,
+                            pid_value=value,
+                            object_type="rec",
+                            object_uuid=record.pid.object_uuid,
+                            status=PIDStatus.RESERVED,
+                        )
+                    except PIDAlreadyExists:
+                        errors.append(already_exists_error_message(scheme, value))
+
+        # Delete the remaining pairs from the database
+        for pid in existing_pairs.values():
+            if pid.object_uuid == record.pid.object_uuid:
+                # Only delete the PIDs not in draft anymore, i.e. targetting this record
+                db.session.delete(pid)
```

Comment:
Great, thanks for the fix above, in the next PR, can you please add something like (pseudocode):
```
assert pid.pid_type not in [recid, doi, oai]
```

It is just a safety check!

Reply:
Added!